Implement new argon osu!taiko hitsounds

peppy opened the pull request:
Second attempt, much more contained than last time.

Quoting @nekodex:

> This PR aims to enhance Taiko hitsounding by adding more variety to in-game sample playback, without diverging too > much from the established or being too distracting. This is achieved through a few different methods, listed roughly > in increasing order of noticability:
> - ~~Subtle pitch randomisation added to hits (to alleviete the 'machine-gun effect' from repeated sample playback)~~
> - Adding (subtle) left/right panning depending on which side the player hits
> - Support different hit/rim samples depending on hitobject volume (by default the thresholds are: 0-59 soft, 60-89 > medium, 90+ hard, but I envision these being changable by mappers). Similar idea to banks, but with less > distinctiveness between each. The goal was to (crudely) simulate how the timbre of real drums change as you hit them harder, allowing for quieter sounding hits during softer sections whilst still being able to belt it out during loud sections.
> - Unique samples for strong hits and strong rim hits - I believe 'stable' osu! layered hitfinish on top of strong hits to do something similar, but this PR replaces that layering (for Argon) and handles the strong playback at input time so players can mash it for fun. This also fixes the issue with strong hits being too loud (due to identical sample being played twice) by cancelling playback of the first sample when a double-hit is detected.
> - Added a 'flourish' sample to strong rim hits. By default these get added automatically to beatmaps (in reverse, to bias towards the end of combos), leaving at least 2s between each. I envision mappers will be able to add these in manually eventually
> 
> Legacy skins should still sound/behave as they did before, apart from balance changes.

To make this work well, I first removed all remaining sample playback logic from `DrawableTaikoHitObject`. The `DrumSamplePlayer` is now solely responsible for hitsound playback.

This allowed implementing completely different logic for argon skin specifically, using the standard skinnable component flow.

- [x] Depends on #24088
- [x] Depends on #24121 
- [x] Depends on https://github.com/ppy/osu/pull/24142

peppy committed:
Move `DrumSamplePlayer` to be a skinnable component

peppy committed:
Add argon-specific `DrumSamplePlayer`

peppy committed:
Add back required pieces to `GameplaySampleTriggerSource` from old PR

peppy committed:
Add balance adjust to base implementation of `DrumSampleTriggerSource`

peppy committed:
Add basic implementation of argon osu!taiko hitsounds (volume / flourish / strong)

peppy committed:
Move strong hit handling to `DrumSamplePlayer` and separte trigger sources

peppy committed:
Change strong hit sample handling to be user input based, not hit object based

peppy committed:
Fix skin fallback not working as expected

peppy committed:
Add time-based flourish support

bdach committed:
Merge branch 'master' into availability-fixes

bdach committed:
Merge pull request #24071 from peppy/availability-fixes

Add handling of "unknown" beatmap availability in multiplayer flow

peppy committed:
Consume nested object states in `HitObjectLifetimeEntry`

peppy committed:
Don't attempt to play drum samples when rewinding

peppy committed:
Warm up pool with argon-specific drum samples

peppy committed:
Fix failing test

peppy committed:
Update resources

peppy committed:
Fix weird test assertion output

peppy committed:
Fix fallback for `Judged` to be more correct

Without this change, when the `Judged` value is checked on an
`HitObjectLifetimeEntry` it would return `true` if a `DrawableHitObject`
has not yet been associated with the entry. Which is completely wrong.

Of note, the usage in `DrawableHitObject` will have never fallen through
to this incorrect value as they always have a result populated:

https://github.com/ppy/osu/blob/f26f001e1d01ca6bb53225da7bf06c0ad21153c5/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L721-L726

nekodex commented:
I'm still going through this, but I noticed that if a specific velocity hit is played (e.g. drum) and you rewind/seek backward, it'll be stuck playing the previous velocity samples (e.g. drum) instead of the new/current ones (e.g. soft). 

Seems to happen both in the editor and during gameplay. Seeking forwards doesn't seem to exhibit the same behaviour, fwiw.

peppy committed:
Add test coverage of `Judged` state

peppy committed:
Merge branch 'fix-judged-fallback' into taiko-hitsounding-final-attempt

peppy committed:
Remove outdated TODO

peppy committed:
Update test assumptions

peppy committed:
Cache `Playfield` for the sake of tests

I'm open to an alternative. Name it.

peppy commented:
> I noticed that if a specific velocity hit is played (e.g. drum) and you rewind/seek backward, it'll be stuck playing the previous velocity samples (e.g. drum) instead of the new/current ones (e.g. soft).

Could you re-test with the latest changes? I think this was caused by https://github.com/ppy/osu/pull/24121 (is already merged into this PR).

nekodex commented:
> Could you re-test with the latest changes? I think this was caused by #24121 (is already merged into this PR).

Hrm, still seems to happen.

peppy commented:
I think we're going to need this:

```diff
diff --git a/osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs b/osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs
index 18d412ab44..8e1b2398da 100644
--- a/osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs
+++ b/osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs
@@ -85,6 +85,14 @@ protected virtual void ApplySampleInfo(SkinnableSound hitSound, ISampleInfo[] sa
                 sound.Stop();
         });
 
+        protected override void Update()
+        {
+            base.Update();
+
+            if (gameplayClock?.ElapsedFrameTime < 0)
+                mostValidObject = null;
+        }
+
         protected HitObject? GetMostValidObject()
         {
             if (mostValidObject == null || isAlreadyHit(mostValidObject))

```

You can test this works (it should) but I'll add proper tests before committing.

nekodex commented:
Noticed a few more minor things:
- Rim hits can incorrectly play as strong in the editor when seeking backwards:

https://github.com/ppy/osu/assets/272140/b53548e8-78ed-4ee0-bca0-6a142ef14b58

- Flourish playback seems to get cutoff by strong rims - i.e. if there's enough strong rim hits within the ~2s window, a flourish may play and then get cut off by a following strong rim hit. Sounds awkward, but not sure if it'll be enough of an issue in practice? (I only stumbled across it while trying to isolate the previous issue)

Besides the above, everything else seems to behave as I had envisioned. 👍

My only concern is the overriding of mappers' bank selections, which may (further) discourage them from properly hitsounding osu!taiko maps... that said, it seems like most aren't bothering anyway? So maybe it's a moot point.

bdach committed:
Merge branch 'master' into taiko-hitsounding-final-attempt

peppy committed:
Expose rewinding state of `IGameplayClock`s

The implementation of this requires a bit of a special case
for 0, so makes sense to implement in a central place.

peppy committed:
Avoid strong hits cutting off other strong hits

peppy commented:
> Flourish playback seems to get cutoff by strong rims - i.e. if there's enough strong rim hits within the ~2s window, a flourish may play and then get cut off by a following strong rim hit. Sounds awkward, but not sure if it'll be enough of an issue in practice? (I only stumbled across it while trying to isolate the previous issue)

I don't think we need to have strong hits cutting each other off in the first place. The main goal was to cut off the non-strong counterparts. I've changed this in the latest commit, see what you think.

peppy committed:
Fix `GameplaySampleTriggerSource` not handling rewinds correctly

peppy commented:
> Rim hits can incorrectly play as strong in the editor when seeking backwards

I can't seem to repro this with the rewind fixes. Can you test again with my rewind changes applied (I've committed a more resilient version now)?

nekodex commented:
> I don't think we need to have strong hits cutting each other off in the first place. The main goal was to cut off the non-strong counterparts. I've changed this in the latest commit, see what you think.

I think it's important the cutting off still happens - the original goal wasn't _just_ to cut off the non-strong counterparts, but also to cut off other strong hits so they'd not compound. Heck, the only reason I was comfortable making the strong hit samples so long (and with rumbly low-frequencies) was because I knew they'd cut each other off... it starts sounding pretty bad once they compound:

https://github.com/ppy/osu/assets/272140/623a20c6-cdcd-4e1c-b0b7-71d297e3958a

That example is arguably the 'best' of worst-case scenarios (loud song & two compounding strongs plus a strong rim) and it's already distorting for me... and this is pretty mild case.

There's also this case of just a bunch of strong hits in a row that sounds pretty bad to me:

https://github.com/ppy/osu/assets/272140/918dc30e-579d-4257-ae6e-1e7a61174426

(The videos don't seem to capture the full extent of the distortion I'm experiencing, but you can test yourself with these beatmaps: https://osu.ppy.sh/beatmapsets/1613520#taiko/3300152 and https://osu.ppy.sh/beatmapsets/83560#taiko/635009)

I'm sure there are other worse examples, but the above were just a few that I could find quickly.


> I can't seem to repro this with the rewind fixes. Can you test again with my rewind changes applied (I've committed a more resilient version now)?

Digging into this again, it may not be connected to an adjacent strong after all? Also seems to happen on a non-argon skin...

https://github.com/ppy/osu/assets/272140/823ad64e-2aad-4d32-820a-3cc16df9f49f

This is the beatmap: https://osu.ppy.sh/beatmapsets/1318794#taiko/2732016 - doesn't seem to happen in master FWIW.

nekodex commented:
Yeah, I think that'd work. Flourishes shouldn't happen more than once every ~2s anyway, so can probably just treat them like any other normal gameplay sample.

peppy committed:
Revert cancelling logic changes based on review feedback

peppy committed:
Move trigger source to own file

Having such a large nested class inside a small top level class is VERY
confusing.

peppy committed:
Move `VolumeAwareHitSampleInfo` to own file

peppy committed:
Move flourish playback to own trigger source

peppy commented:
I've fixed the flourish cancellation issue. It wasn't as simple as I hoped due to the logic deciding whether to play flourishes being one level too low. @bdach curious if you have any better suggestions than the approach I took. It's not *that* bad, but I'd hope we could avoid doing the whole `GetMostValidObject()` twice, maybe.

peppy committed:
Implement flourish trigger source via base class

peppy committed:
Fix rewinding causing incorrectly stronged non-strong hits

peppy commented:
I've fixed the editor rewind issue 🤦‍♂️

This should be... ready to go again.

nekodex commented on osu.Game.Rulesets.Taiko/UI/DrumSamplePlayer.cs:
```diff
@@ -1,57 +1,150 @@
+        {
+            leftRimTrigger.StopAllPlayback();
+            rightRimTrigger.StopAllPlayback();
+            strongCentreTrigger.StopAllPlayback();
```
```suggestion
            strongRimTrigger.StopAllPlayback();
```

peppy committed:
Update osu.Game.Rulesets.Taiko/UI/DrumSamplePlayer.cs

Co-authored-by: Jamie Taylor <me@nekodex.net>

bdach committed:
Merge branch 'master' into taiko-hitsounding-final-attempt

bdach commented:
> It's not _that_ bad, but I'd hope we could avoid doing the whole `GetMostValidObject()` twice, maybe.

I'm not entirely sure what this is referring to because I don't see anything `GetMostValidObject()` twice.

Is this a mental shortcut following from the fact that flourish has been split off to its own trigger source? In that case I'm not sure I have anything better to offer and what you have is probably fine to go with until proven otherwise.

bdach committed:
Use new `IGameplayClock.IsRewinding` member

peppy commented:
> Is this a mental shortcut following from the fact that flourish has been split off to its own trigger source? In that case I'm not sure I have anything better to offer and what you have is probably fine to go with until proven otherwise.

Yeah, it's just the fact one extra trigger source is doing the logic. But I don't think it's of a huge concern, just something that felt a bit weird.

FWIW nekodex has confirmed he has no remaining issues with this, to his knowledge.

bdach committed:
Fix flourish sample not playing if strong hits are hit early

bdach commented:
I've done a.. moderate? amount of testing of this. Out of that testing fell out https://github.com/ppy/osu/pull/24119/commits/ba0cd7a3f27206e5b815b709da91bf46969e3c90.

Other than that, one small post-https://github.com/ppy/osu/pull/24142 cleanup in https://github.com/ppy/osu/pull/24119/commits/fee56ac6d2ece7d897a6419d0bfbc9ee65e95c19. Other than _that_, probably good to go. In my book there's a 50/50 chance this series of changes broke something, but (a) I would estimate that the original version of this had an 80/20 chance of breaking something, and (b) we've been trying to fit this thing for so long now that I'm just going to accept this as-is and attempt to roll with the punches if anything does indeed break.

Egaelytra commented:
Do you make musical introduction "Redside - Beachballs" on laser?
